### PR TITLE
feat(flow): Update locked wallet state in flash

### PR DESCRIPTION
**Pull request by amanCypherock**
https://app.clickup.com/t/9002019994/PRF-4840

This PR handles the locked wallet case in reconstructing wallet flow.
- Wallet-locked status is updated in Flash along with the card number.
- Fixes remaining cards update in `card_fetch_share`
- Fixes `get_wallet_data_by_id` and `get_wallet_name_by_id` to return false if the wallet is locked.

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-4840 Locked wallet handling and UX](https://app.clickup.com/t/9002019994/PRF-4840)

File: common/interfaces/flash_interface/flash_api.h [hunk 1]
```diff
@@ -297,6 +297,20 @@ int update_challenge_flash(const char *name,
...
+ *
+ * @param wallet_id The wallet_id is used to determine which wallet is locked.
+ * If that wallet is found, it's status is updated to locked
+ * @param card_locked It is used to determine on which card is the wallet
+ * locked. The value of "card_locked" should be between 1 and 4, inclusive.
+ *
+ * @return SUCCESS if execution is successful, error is returned if any error is
+ * encountered duirng wallet fetching.
+ */
+int set_wallet_locked(const char *wallet_name, uint8_t card_locked);
```
Comment (ujjwal-cyph) [inline common/interfaces/flash_interface/flash_api.h:312]: the 2nd param name does not fit for readability. Suggestion:

`card_locked` -> `card_number`

File: src/wallet/reconstruct_seed_flow.c [hunk 2]
```diff
@@ -153,10 +153,10 @@ reconstruct_state_e reconstruct_seed_handler(reconstruct_state_e state,
       ui_scrollable_page(
           ui_text_confirm_passphrase, display, MENU_SCROLL_HORIZONTAL, false);
       memzero(display, sizeof(display));
-      next_state = get_state_on_confirm_scr(
-          PASSPHRASE_CONFIRM, PASSPHRASE_INPUT, TIMED_OUT);
+      next_state =
+          get_state_on_confirm_scr(PIN_INPUT, PASSPHRASE_INPUT, TIMED_OUT);
 
-      if (PASSPHRASE_CONFIRM == next_state) {
+      if (PIN_INPUT == next_state) {
```
Comment (ujjwal-cyph) [inline src/wallet/reconstruct_seed_flow.c:159]: would be best if we split this fix as a separate PR

File: common/interfaces/flash_interface/flash_api.h [hunk 3]
```diff
@@ -297,6 +297,20 @@ int update_challenge_flash(const char *name,
                            const uint8_t target[SHA256_SIZE],
                            const uint8_t random_number[POW_RAND_NUMBER_SIZE]);
 
+/**
+ * The function sets the wallet locked status and updates the locked card
+ * number(1-4).
+ *
+ * @param wallet_id The wallet_id is used to determine which wallet is locked.
+ * If that wallet is found, it's status is updated to locked
```
Comment (kartikCypherock) [inline common/interfaces/flash_interface/flash_api.h:305]:
```suggestion
```

Comment (kartikCypherock) [inline common/interfaces/flash_interface/flash_api.h:305]: Please update this comment. We are using wallet_name as key

File: src/card_flows/card_flow_reconstruct_wallet.c [hunk 4]
```diff
@@ -97,12 +98,23 @@
...
 
-  return card_fetch_share(&configuration);
+  if (CARD_OPERATION_LOCKED_WALLET == card_status) {
+    int status = set_wallet_locked(
+        (const char *)wallet.wallet_name,
+        decode_card_number(config.remaining_cards ^ acceptable_cards));
+
+    if (SUCCESS != status) {
+      LOG_CRITICAL("xxx38: %d", status);
+    }
```
Comment (kartikCypherock) [inline src/card_flows/card_flow_reconstruct_wallet.c:117]: Canwe please move this to card operation instead? We are handling WRONG_PIN there as well

File: src/wallet/reconstruct_seed_flow.c (hunk 2 above)
Comment (amanCypherock) [reply src/wallet/reconstruct_seed_flow.c:159]: raised separate [PR](https://github.com/Cypherock/x1_wallet_firmware/pull/334)

File: common/interfaces/flash_interface/flash_api.h (hunk 1 above)
Comment (amanCypherock) [reply common/interfaces/flash_interface/flash_api.h]: updated

File: common/interfaces/flash_interface/flash_api.h (hunk 3 above)
Comment (amanCypherock) [reply common/interfaces/flash_interface/flash_api.h:305]: updated

File: src/card_flows/card_flow_reconstruct_wallet.c (hunk 4 above)
Comment (amanCypherock) [reply src/card_flows/card_flow_reconstruct_wallet.c]: updated

File: common/interfaces/flash_interface/flash_api.h (hunk 3 above)
Comment (kartikCypherock) [reply common/interfaces/flash_interface/flash_api.h:305]: I think doxygen comments are duplicated for some reason, please check and fix

Comment (amanCypherock) [reply common/interfaces/flash_interface/flash_api.h]: updated

File: src/card_operations/card_utils.c [hunk 5]
```diff
@@ -165,16 +165,28 @@ card_error_type_e wait_for_card_removal(void) {
...
+      return indicate_card_error(error_message);
+    }
+  } else if (CARD_OPERATION_LOCKED_WALLET == card_data->error_type) {
+    int status =
+        set_wallet_locked((const char *)wallet->wallet_name,
+                          decode_card_number(card_data->nfc_data.tapped_card));
+
+    if (SUCCESS != status) {
+      LOG_CRITICAL("xxx38: %d", status);
+    }
```
[the remark this answers is not in the document]
Comment (amanCypherock) [reply src/card_operations/card_utils.c:188]: We would have to be consistent with the `card_error_type_e` so can't return the error from flash API, will add a return code, so it is caught though.

[the remark this answers is not in the document]
Comment (amanCypherock) [reply src/card_operations/card_utils.c:188]: We are not looking for P0 events here and any error we return from here will override the locked wallet error.

I'm unsure on how else should we handle something not updated on flash as the application will neither have the context or enough data to resolve or fix this.

I think a log would be enough in this case.

[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply src/card_operations/card_utils.c:188]: okay. i was concerned if the card error is shadowed by line 191 as the function returns `CARD_OPERATION_SUCCESS`.  but its not the case.

i got confused. resolving

**Review by kartikCypherock — APPROVED** [review]
